### PR TITLE
[Merged by Bors] - feat(Archive/Imo): IMO 1963 Q5

### PR DESCRIPTION
**Pull request by Rida-Hamadani**
proves the equality `cos (π / 7) - cos (2 * π / 7) + cos (3 * π / 7) = 1 / 2`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Archive/Imo/Imo1963Q5.lean [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  intro k
+  ring_nf
+  by_contra h
+  have seven_ne_zero : (7 : ℝ) ≠ 0 := by simp
+  rw [mul_comm, mul_right_inj' pi_ne_zero, ← mul_right_inj' seven_ne_zero] at h
+  simp only [one_div, isUnit_iff_ne_zero, ne_eq, OfNat.ofNat_ne_zero, not_false_eq_true,
+    IsUnit.mul_inv_cancel] at h
+  norm_cast at h
+  apply Int.eq_one_of_mul_eq_one_right (by decide) at h
+  contradiction
```
Comment (Command-Master) [inline Archive/Imo/Imo1963Q5.lean:35]:
```suggestion
lemma two_sin_pi_over_seven_ne_zero : 2 * sin (π / 7) ≠ 0 := by
  simp only [ne_eq, mul_eq_zero, OfNat.ofNat_ne_zero, false_or]
  apply (Real.sin_pos_of_pos_of_lt_pi _ _).ne' <;> linarith [pi_pos]
```

File: Archive/Imo/Imo1963Q5.lean [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+Authors: Rida Hamadani
+-/
+import Mathlib.Analysis.SpecialFunctions.Trigonometric.Complex
+
+/-!
+# IMO 1963 Q5
+
+Prove that `cos (π / 7) - cos (2 * π / 7) + cos (3 * π / 7) = 1 / 2`.
+
+The main idea of the proof is multiply both sides by `2 * sin (π / 7)`, then the result follows
```
Comment (grunweg) [inline Archive/Imo/Imo1963Q5.lean:13]:
```suggestion
The main idea of the proof is to multiply both sides by `2 * sin (π / 7)`, then the result follows

File: Archive/Imo/Imo1963Q5.lean [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+open Real
+
+lemma cos_times_sin (x y : ℝ) : 2 * sin x * cos y = (sin (x + y) + sin (x - y)) := by
+  simp [sin_add, sin_sub]
+  ring
+
+lemma two_sin_pi_over_seven_ne_zero : 2 * sin (π / 7) ≠ 0 := by
+  simp only [ne_eq, mul_eq_zero, OfNat.ofNat_ne_zero, false_or]
+  apply (Real.sin_pos_of_pos_of_lt_pi _ _).ne' <;> linarith [pi_pos]
```
Comment (vihdzp) [inline Archive/Imo/Imo1963Q5.lean:26]:
```suggestion
  apply mul_ne_zero two_ne_zero (Real.sin_pos_of_pos_of_lt_pi _ _).ne' <;> linarith [pi_pos]
```

**Review by vihdzp — COMMENTED** [review]
Otherwise LGTM. Quite surprising to see the IMO had trig problems at some point.

File: Archive/Imo/Imo1963Q5.lean [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+Prove that `cos (π / 7) - cos (2 * π / 7) + cos (3 * π / 7) = 1 / 2`.
+
+The main idea of the proof is to multiply both sides by `2 * sin (π / 7)`, then the result follows
+through basic algebraic manipulations with the use of some trigonometric identities.
+
+-/
+
+open Real
+
+lemma cos_times_sin (x y : ℝ) : 2 * sin x * cos y = (sin (x + y) + sin (x - y)) := by
```
[the remark this answers is not in the document]
Comment (Rida-Hamadani) [reply Archive/Imo/Imo1963Q5.lean]: Done. Please let me know if you think that there is a more appropriate place for the lemmas, as I am not very familiar with this side of mathlib.

File: Archive/Imo/Imo1963Q5.lean [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+Prove that `cos (π / 7) - cos (2 * π / 7) + cos (3 * π / 7) = 1 / 2`.
+
+The main idea of the proof is to multiply both sides by `2 * sin (π / 7)`, then the result follows
+through basic algebraic manipulations with the use of some trigonometric identities.
+
+-/
+
+open Real
+
+lemma two_sin_pi_over_seven_ne_zero : 2 * sin (π / 7) ≠ 0 := by
```
Comment (jcommelin) [inline Archive/Imo/Imo1963Q5.lean:20]:
```suggestion
lemma two_sin_pi_div_seven_ne_zero : 2 * sin (π / 7) ≠ 0 := by
```

File: Archive/Imo/Imo1963Q5.lean [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  apply mul_ne_zero two_ne_zero (Real.sin_pos_of_pos_of_lt_pi _ _).ne' <;> linarith [pi_pos]
+
+lemma sin_pi_mul_neg_div (a b : ℝ) : sin (π * (- a / b)) = - sin (π * (a / b)) := by
+  ring_nf
+  exact sin_neg _
+
+theorem imo1963_q5 : cos (π / 7) - cos (2 * π / 7) + cos (3 * π / 7) = 1 / 2 := by
+  rw [← mul_right_inj' two_sin_pi_over_seven_ne_zero, mul_add, mul_sub, ← sin_two_mul,
```
Comment (jcommelin) [inline Archive/Imo/Imo1963Q5.lean:28]:
```suggestion
  rw [← mul_right_inj' two_sin_pi_div_seven_ne_zero, mul_add, mul_sub, ← sin_two_mul,
```

**Review by jcommelin — COMMENTED** [review]
bors d+

**Pull request closed without merging**